File: programmers/new/level2/주식가격.java
import java.io.*;
import java.util.*;
class Node{
	int idx;
	int number;
	Node(int idx, int number){
		this.idx = idx;
		this.number = number;
	}
}
class Solution {
    public int[] solution(int[] prices) {
        int[] answer = new int[prices.length];
        Stack<Node> stack = new Stack<Node>();
        for(int i = 0; i < answer.length; ++i) {
        	if(stack.isEmpty()) {
        		stack.push(new Node(i, prices[i]));
        	}else {
        		while(!stack.isEmpty() && prices[i]<stack.peek().number) {
        			Node node = stack.pop();
        			answer[node.idx] = i-node.idx;
        		}
        		stack.push(new Node(i, prices[i]));
        	}
        }
        while(!stack.isEmpty()) {
			Node node = stack.pop();
			answer[node.idx] = answer.length-1-node.idx;
		}
        return answer;
    }
}
public class 주식가격 {
	public static void main(String[] args) throws Exception{
		int[] prices = {1,2,3,2,3};
		Solution sol = new Solution();
		System.out.println(Arrays.toString(sol.solution(prices)));
	}
}

File: programmers/python/level2/주식가격.py
from collections import deque
def solution(prices):
    st = deque()
    answer = [0 for i in range(len(prices))]
    for i, p in enumerate(prices):
        if len(st) == 0:
            st.append({'index': i, 'value': p})
        else:
            while len(st) > 0 and st[-1]['value'] > p:
                top = st.pop()
                answer[top['index']] = i - top['index']
            st.append({'index': i, 'value': p})
    while len(st) > 0:
        top = st.pop()
        answer[top['index']] = len(answer) - 1 - top['index']
    return answer

print(solution([1,2,3,2,3]))